L74-java-channel-creds.md

```diff
@@ -0,0 +1,485 @@
+users typically have the security information available before creating the
+builder and reusing a builder is rare, especially to change the security. 
+
+```java
+package io.grpc;
+
+/**
+ * Represents a security configuration to be used for channels. There is no
+ * generic mechanism for processing arbitrary ChannelCredentials; the consumer
+ * of the credential (the channel) must support each implementation explicitly
+ * and separately. Consumers are not required to support all types or even all
+ * possible configurations for types that are partially supported, but they
+ * <em>must</em> at least fully support ChoiceChannelCredentials.
+ *
+ * <p>A Credential provides client identity and authenticates the server. This
+ * is different from CallCredentials, which only provides client identity. They
+ * can also influence types of encryption used and similar security
+ * configuration.
+ */
+public abstract class ChannelCredentials {}
+
+public final class TlsChannelCredentials extends ChannelCredentials {
+  /* Complicated. Described separately below. */
+}
+
+/** No client identity, authentication, or encryption is to be used. */
+public final class InsecureChannelCredentials extends ChannelCredentials {}
+
+/**
+  * Provides a list of ChoiceChannelCredentials, where any one may be used. The
+  * credentials are in preference order.
+  */
+public final class ChoiceChannelCredentials extends ChannelCredentials {
+  public ChoiceChannelCredentials(ChannelCredentials... creds) {...}
+
+  public List<ChannelCredentials> getCredentialsList() { return creds; }
+}
+```
+
+`ChoiceChannelCredentials` is intended to be used in cases like
+`GoogleDefaultCredentials`, where TLS may be satisfactory, but there may also
+be more optimal credentials if the transport is a certain type.
+
+To allow choosing `ChannelCredentials` and `CallCredentials` simultaneously,
+there will be a composite credential. If `CallCredentials` are specified on the
+Channel and per-RPC, then _both_ `CallCredentials` will be used. This matches
+the behavior in other languages.
+
+```java
+package io.grpc;
+
+public final class CompositeChannelCredentials extends ChannelCredentials {
+  public CompositeChannelCredentials(
+    ChannelCredentials chanCreds, CallCredentials callCreds) {...}
+  public ChannelCredentials getChannelCredentials() { return channelCredentials; }
+  public CallCredentials getCallCredentials() { return callCredentials; }
+}
+```
+
+Unlike most of the credentials, TLS may have a considerable amount of
+configuration. And as time goes on, more configuration options may be provided.
+It could be a severe bug if certain TLS restrictions were ignored by a channel.
+To detect cases where the channel does not understand enough of the TLS
+configuration, there will be a method to check if there are required features
+that are not understood. If some are not understood, enough information will be
```

Comment:
Just to confirm my understanding since I haven't seen this pattern before, the way this is going to work is as follows: 

A transport implementation such as Netty will call `incomprehensible()` with the set of `Feature`s it (Netty) supports and if the `TlsChannelCredentials` instance returns a non-empty set then there is something Netty doesn't support (that `TlsChannelCredentials` requires) and it will throw an error (with error text containing non-supported feature set).

So this pattern depends on the transport implementation verifying features thru `incomprehensible()` to prevent  a required TLS restriction/feature from being ignored by the transport.

Reply:
> and if the TlsChannelCredentials instance returns a non-empty set then there is something Netty doesn't support (that TlsChannelCredentials requires) and it will throw an error (with error text containing non-supported feature set).

If "it" means "Netty," yes.

> So this pattern depends on the transport implementation verifying features thru incomprehensible() to prevent a required TLS restriction/feature from being ignored by the transport.

Yes.